### PR TITLE
Add canceler to `Task.all`, and tests. Issue #260.

### PR DESCRIPTION
**Pull request by jacobroufa**

### CODE REVIEW

**Comment by jacobroufa** [comment]
It looks like the Travis failures above are for an `async` test in Firefox and the `text` plugin on Android -- unrelated to the `Task.all` canceler. Is there anything I should do in this case?

**Comment by dylans** [comment]
Per discussion, let's squash and merge this as the CI failure is unrelated to these changes.

**Review by dylans — APPROVED** [review]

File: src/async/Task.ts [hunk 1]
```diff
@@ -72,7 +72,28 @@ export default class Task<T> extends ExtensiblePromise<T> {
...
+		}, () => {
+			if (isIterable(iterable) || isArrayLike(iterable)) {
+				forOf(iterable, (promiseLike: any) => {
+					if (isTask(promiseLike)) {
+						promiseLike.cancel();
+					}
+				});
+			}
+			else {
+				const promiseKeys = Object.keys(iterable);
```
Comment (jason0x43) [inline src/async/Task.ts]: Is a separate keys constant necessary?

File: src/async/Task.ts [hunk 2]
```diff
@@ -72,7 +72,28 @@ export default class Task<T> extends ExtensiblePromise<T> {
 	static all<T>(iterable: T | Thenable<T>): Task<T[]>;
 	static all<T>(iterable: ListOfPromises<T>): Task<T[]>;
 	static all<T>(iterable: DictionaryOfPromises<T> | ListOfPromises<T>): Task<any> {
-		return <Task<T>> super.all(iterable);
+		return new Task<T>((resolve, reject) => {
+			super.all(iterable).then(<any> resolve, <any> reject);
+		}, () => {
+			if (isIterable(iterable) || isArrayLike(iterable)) {
+				forOf(iterable, (promiseLike: any) => {
+					if (isTask(promiseLike)) {
```
Comment (agubler) [inline src/async/Task.ts]: missing coverage for a branch

File: src/async/Task.ts [hunk 3]
```diff
@@ -72,7 +72,28 @@ export default class Task<T> extends ExtensiblePromise<T> {
...
+					}
+				});
+			}
+			else {
+				const promiseKeys = Object.keys(iterable);
+
+				promiseKeys.forEach((key: any) => {
+					const promiseLike = iterable[ key ];
+
+					if (isTask(promiseLike)) {
```
Comment (agubler) [inline src/async/Task.ts]: Missing coverage for a branch

File: tests/unit/async/Task.ts [hunk 4]
```diff
@@ -372,6 +372,46 @@ function addPromiseTests(suite: any, Promise: any) {
...
+
+				for (let i = 0; i < 3; i++) {
+					pending[i] = new Promise(function () {});
+				}
+
+				let tasks = Promise.all(pending);
+
+				tasks.cancel();
+
+				tasks.then(() => {
```
Comment (jason0x43) [inline tests/unit/async/Task.ts]: The test should also verify that the pending tasks were canceled.

File: tests/unit/async/Task.ts [hunk 5]
```diff
@@ -372,6 +372,46 @@ function addPromiseTests(suite: any, Promise: any) {
...
 				assert.notStrictEqual(value, iterable);
 			}));
+		},
+
+		'cancelable': {
+			'isIterable': function (this: any) {
+				let pending: any[] = [];
+
+				for (let i = 0; i < 3; i++) {
+					pending[i] = new Promise(function () {});
```
Comment (jason0x43) [inline tests/unit/async/Task.ts]: Some/all of the pending things should be Tasks.

File: src/async/Task.ts [hunk 6]
```diff
@@ -72,7 +72,28 @@ export default class Task<T> extends ExtensiblePromise<T> {
 	static all<T>(iterable: T | Thenable<T>): Task<T[]>;
 	static all<T>(iterable: ListOfPromises<T>): Task<T[]>;
 	static all<T>(iterable: DictionaryOfPromises<T> | ListOfPromises<T>): Task<any> {
-		return <Task<T>> super.all(iterable);
+		return new Task<T>((resolve, reject) => {
+			super.all(iterable).then(<any> resolve, <any> reject);
```
Comment (jason0x43) [inline src/async/Task.ts]: It seems odd that we'd have to cast the resolve and reject callbacks.

**Comment by dylans** [comment]
We should decide what to do with this one as it is getting stale.

**Comment by kitsonk** [comment]
IMO, Jason's comments appear to be valid and should be addressed.

File: src/async/Task.ts (hunk 6 above)
Comment (jacobroufa) [reply src/async/Task.ts]: If I don't cast them, TSC throws this error:

```
src/async/Task.ts(76,29): error TS2345: Argument of type '(value?: T | Thenable<T> | undefined) => void' is not assignable to parameter of type '((value: (DictionaryOfPromises<T>
| Iterable<T | Thenable<T>>)[]) => void | Thenable<void> | unde...'.
  Type '(value?: T | Thenable<T> | undefined) => void' is not assignable to type '(value:(DictionaryOfPromises<T> | Iterable<T | Thenable<T>>)[]) => void | Thenable<void> | undef.
..'.
    Types of parameters 'value' and 'value' are incompatible.
      Type '(DictionaryOfPromises<T> | Iterable<T | Thenable<T>>)[]' is not assignable totype 'T | Thenable<T> | undefined'.
        Type '(DictionaryOfPromises<T> | Iterable<T | Thenable<T>>)[]' is not assignable to type 'Thenable<T>'.
          Property 'then' is missing in type '(DictionaryOfPromises<T> | Iterable<T | Thenable<T>>)[]'.
```

File: src/async/Task.ts (hunk 1 above)
Comment (jacobroufa) [reply src/async/Task.ts]: Addressed in 39fe3eb.

File: src/async/Task.ts (hunk 2 above)
Comment (jacobroufa) [reply src/async/Task.ts]: Addressed in 39fe3eb.

File: src/async/Task.ts (hunk 3 above)
Comment (jacobroufa) [reply src/async/Task.ts]: Addressed in 39fe3eb.

File: tests/unit/async/Task.ts (hunk 4 above)
Comment (jacobroufa) [reply tests/unit/async/Task.ts]: This is literally what the tests are asserting: `assert.strictEqual(task.state, State.canceled, 'Task should have Canceled state');` where `task` is each item of `pending`.

File: tests/unit/async/Task.ts (hunk 5 above)
Comment (jacobroufa) [reply tests/unit/async/Task.ts]: Addressed in 39fe3eb.

**Comment by jacobroufa** [comment]
Feedback addressed in 39fe3eb. Please let me know if there's anything else on this one!

**Comment by kitsonk** [comment]
@maier49 we need to do whatever with this PR, either get it to pass the tests, resolve conflicts, whatever...

**Comment by maier49** [comment]
@kitsonk I resolved the conflicts and resolved the last piece of feedback from Jason that wasn't already addressed. The build is still failing but that should be unrelated since master is failing and all of the task tests pass.

**Review by kitsonk — APPROVED** [review]
Agreed the CI Failure is unreleated.

**Pull request merged**